### PR TITLE
handling IPv6 within square brackets

### LINKED ISSUE #32978: IPv6 address is not correctly parsed by filebeat with the nginx_ingress_controller module
IPv6 addresses can be surrounded by `[]` like `[2a02:cf40::]:9200` in some nginx ingress controller logs

```
2a02:cf40:: - remote_monitoring_user [24/Aug/2022:21:04:17 +0000] "POST /_bulk HTTP/1.1" 200 470 "-" "Elastic-metricbeat/7.16.3 (linux; amd64; e7cede6a62ed4452bd9044fc6f4947df; 2022-01-07 00:50:33 +0000 UTC)" 2057 0.033 [esmon-esmon-es-http-9200] [] [2a02:cf40::]:9200 470 0.036 200 3db73c6c673c4256ade033a6ce08c2ab
```

In some logs you can even find a list of ipv6/port like `[2a02:cf40::7]:5000, [2a02:cf40::4e36]:5000`

```
2a02:cf40::4e36 - - [24/Aug/2022:18:05:41 +0000] \"GET /favicon.ico HTTP/2.0\" 502 552 \"https://localhost:8080/\" \"Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/104.0.0.0 Safari/537.36\" 27 0.000 [localhost-8080] [] [2a02:cf40::7]:5000, [2a02:cf40::4e36]:5000 0, 0, 0 0.000, 0.000, 0.000 502, 502, 502 3db73c6c673c4256ade033a6ce08c2ab
```

The grok pattern in the ingest_pipeline `beats/filebeat/module/nginx/ingress_controller/ingest/pipeline.yml` has to be modified to correctly handle this use case. Furthermore we need to correctly split the ip address from the port.

I found some references to the optional square brackets surrounding the IPv6 [here](https://serverfault.com/questions/1026466/ipv6-address-text-notation-with-prefix-inside-or-outside-square-brackets). This article points to an official RFC suggesting that IPv6 should be surrounded by square brackets in URLs to distinguish IPv6 from ports. 

- Version: 8.4
- Operating System: Any
- Steps to Reproduce: 

try to parse the provided log lines with the ingest pipeline provided by `beats/filebeat/module/nginx/ingress_controller/ingest/pipeline.yml`

### PR DESCRIPTION
**Pull request by gsantoro**
## What does this PR do?

Handle [] around IPv6 in upstream servers addresses

## Why is it important?

If upstream server field has an IPv6 it will not be correctly handled. 

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Author's Checklist

- [ ] all pipeline tests are passing. You can run the pipeline tests with

```bash
cd filebeat
GENERATE=true TESTING_FILEBEAT_MODULES=nginx TESTING_FILEBEAT_FILESETS=ingress_controller mage -v pythonIntegTest
```

## Related issues

- Closes #32978

### CODE REVIEW

**Comment by andrewkroh** [comment]
Relates: https://github.com/elastic/integrations/pull/3117#discussion_r855626013

**Comment by gsantoro** [comment]
> Please add test data for the use case that this fixes. Add the samples to https://github.com/elastic/beats/blob/980972ea9a58bf6bb78fb190b499fbf9c17f1cc4/filebeat/module/nginx/ingress_controller/test/test.log then regenerate the expected.json file.

I was looking into adding some tests but currently I can't find how to run those ingest pipeilne tests in the nginx_ingress_controller fileset. Is there something like what we use in the integrations repo (eg. `elastic-package test pipeline -v`).

File: filebeat/module/nginx/ingress_controller/ingest/pipeline.yml [hunk 1]
```diff
@@ -139,18 +140,36 @@ processors:
...
+        - "^%{IPV6NOCOMPRESS:nginx.ingress_controller.upstream.ip}:%{NUMBER:nginx.ingress_controller.upstream.port}$"
+        - "^%{IPV6:nginx.ingress_controller.upstream.ip}%{IPV6PORTSEP}%{NUMBER:nginx.ingress_controller.upstream.port}$"
+        - "^%{IPV6:nginx.ingress_controller.upstream.ip}%{IPV6PORTSEP}%{POSINT:nginx.ingress_controller.upstream.port}$"
+      pattern_definitions:
+        IPV6NOCOMPRESS: '([0-9A-Fa-f]{1,4}:){7}[0-9A-Fa-f]{1,4}'
+        IPV6PORTSEP: '(?: port |[p#.])'
+      ignore_missing: true
+      ignore_failure: true
+  - convert:
+      field: nginx.ingress_controller.upstream.ip
```
Comment (andrewkroh) [inline filebeat/module/nginx/ingress_controller/ingest/pipeline.yml:163]: If the `convert` fails then you should handle it by removing the invalid IP from this field via `on_failure`. If you leave the bad value in a field mapped as an IP then the event will fail to index unless the mapping for the field includes `ignore_malformed: true`.

Same goes for the upstream.port.

Comment (gsantoro) [reply filebeat/module/nginx/ingress_controller/ingest/pipeline.yml:163]: I fixed the pipeline now and added some tests

**Review by tetianakravchenko — APPROVED** [review]
LGTM!

File: filebeat/module/nginx/ingress_controller/ingest/pipeline.yml [hunk 2]
```diff
@@ -139,18 +140,42 @@ processors:
...
+        - "^%{IPV4:nginx.ingress_controller.upstream.ip}:%{NUMBER:nginx.ingress_controller.upstream.port}$"
+        - "^\\[%{IPV6:nginx.ingress_controller.upstream.ip}\\]:%{NUMBER:nginx.ingress_controller.upstream.port}$"
+        - "^%{IPV6NOCOMPRESS:nginx.ingress_controller.upstream.ip}:%{NUMBER:nginx.ingress_controller.upstream.port}$"
+        - "^%{IPV6:nginx.ingress_controller.upstream.ip}%{IPV6PORTSEP}%{NUMBER:nginx.ingress_controller.upstream.port}$"
+        - "^%{IPV6:nginx.ingress_controller.upstream.ip}%{IPV6PORTSEP}%{POSINT:nginx.ingress_controller.upstream.port}$"
+      pattern_definitions:
+        IPV6NOCOMPRESS: '([0-9A-Fa-f]{1,4}:){7}[0-9A-Fa-f]{1,4}'
+        IPV6PORTSEP: '(?: port |[p#.])'
+      ignore_missing: true
+      ignore_failure: true
```
Comment (gizas) [inline filebeat/module/nginx/ingress_controller/ingest/pipeline.yml:161]: So wondering if this has to be ignore_failure: false ?

**Comment by gsantoro** [comment]
@Mergifyio backport 8.3

File: filebeat/module/nginx/ingress_controller/ingest/pipeline.yml (hunk 2 above)
Comment (gsantoro) [reply filebeat/module/nginx/ingress_controller/ingest/pipeline.yml:161]: just to be on the safe side

**Comment by Strangerxxx** [comment]
@Mergifyio backport 7.17

**Comment by Strangerxxx** [comment]
@gsantoro is it possible to backport this into 7.17 branch?

**Pull request merged**